app/code/Magento/CatalogUrlRewrite/Observer/ProductProcessUrlRewriteSavingObserver.php

```
@@ -3,73 +3,168 @@
      */
-    private $productUrlRewriteGenerator;
+    private $urlPersist;
 
     /**
-     * @var UrlPersistInterface
+     * @var StoreWebsiteRelationInterface|null
      */
-    private $urlPersist;
+    private $storeWebsiteRelation;
+
+    /**
+     * @var AppendUrlRewritesToProducts
+     */
+    private $appendRewrites;
 
     /**
-     * @var ProductUrlPathGenerator
+     * @var ScopeConfigInterface
      */
-    private $productUrlPathGenerator;
+    private $scopeConfig;
 
     /**
-     * @param ProductUrlRewriteGenerator $productUrlRewriteGenerator
      * @param UrlPersistInterface $urlPersist
-     * @param ProductUrlPathGenerator|null $productUrlPathGenerator
+     * @param StoreWebsiteRelationInterface|null $storeWebsiteRelation
+     * @param AppendUrlRewritesToProducts|null $appendRewrites
+     * @param ScopeConfigInterface $scopeConfig
      */
     public function __construct(
-        ProductUrlRewriteGenerator $productUrlRewriteGenerator,
         UrlPersistInterface $urlPersist,
-        ProductUrlPathGenerator $productUrlPathGenerator = null
+        StoreWebsiteRelationInterface $storeWebsiteRelation,
+        AppendUrlRewritesToProducts $appendRewrites,
+        ScopeConfigInterface $scopeConfig
     ) {
-        $this->productUrlRewriteGenerator = $productUrlRewriteGenerator;
         $this->urlPersist = $urlPersist;
-        $this->productUrlPathGenerator = $productUrlPathGenerator ?: ObjectManager::getInstance()
-            ->get(ProductUrlPathGenerator::class);
+        $this->storeWebsiteRelation = $storeWebsiteRelation;
+        $this->appendRewrites = $appendRewrites;
+        $this->scopeConfig = $scopeConfig;
     }
 
     /**
      * Generate urls for UrlRewrite and save it in storage
      *
-     * @param \Magento\Framework\Event\Observer $observer
+     * @param Observer $observer
      * @return void
-     * @throws \Magento\UrlRewrite\Model\Exception\UrlAlreadyExistsException
+     * @throws UrlAlreadyExistsException
```

You should not add Exception to annotation if your method does not throw exception directly

---

required by https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html#5-exceptions in 5.3

---

@engcom-Golf please refer to docblock magento standards
https://devdocs.magento.com/guides/v2.3/coding-standards/docblock-standard-general.html#functions-methods
A declaration of possible exceptions using the @throws tag, **if the actual body of function triggers an exception**.

---

@engcom-Golf please refer to docblock magento standards
https://devdocs.magento.com/guides/v2.3/coding-standards/docblock-standard-general.html#functions-methods
A declaration of possible exceptions using the @throws tag, if the actual body of function triggers an exception.

---

Some kind of collision, but I'd like to not hide the possible exception throwing. This was required by @engcom-Kilo review as well. 
https://github.com/magento/magento2/pull/31106#discussion_r535227589